File: Ahorn/Entities/shroomDashSwitch.jl
module ShroomHelperShroomDashSwitch

using ..Ahorn, Maple

sides = String[
    "left",
    "right",
    "up",
    "down"
]

textures = String[
    "default",
    "mirror"
]

@mapdef Entity "ShroomHelper/ShroomDashSwitch" ShroomDashSwitch(
    x::Integer, 
    y::Integer, 
    side::String="up",
    persistent::Bool=false,
    refillDashOnCollision::Bool=false,
    doubleDashRefill::Bool=false,
    isWindTrigger::Bool=false,
    windPatternOnCollision::String="None"
)

const placements = Ahorn.PlacementDict(
    "Shroom Dash Switch ($(uppercasefirst(side))) (Shroom Helper)" => Ahorn.EntityPlacement(
        ShroomDashSwitch,
        "rectangle",
        Dict{String, Any}(
            "side" => side
        )
    ) for side in sides
)

Ahorn.editingOptions(entity::ShroomDashSwitch) = Dict{String, Any}(
    "side" => sides,
    "windPatternOnCollision" => Maple.wind_patterns
)

function Ahorn.selection(entity::ShroomDashSwitch)
    x, y = Ahorn.position(entity)
    side = get(entity.data, "side", false)

    if side == "left"
        return Ahorn.Rectangle(x, y - 1, 10, 16)
    elseif side == "right"
        return Ahorn.Rectangle(x - 2, y, 10, 16)
    elseif side == "down"
        return Ahorn.Rectangle(x, y, 16, 12)
    elseif side == "up"
        return Ahorn.Rectangle(x, y - 4, 16, 12)
    end
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::ShroomDashSwitch, room::Maple.Room)
    sprite = get(entity.data, "sprite", "default")
    side = get(entity.data, "side", "up")
    texture = "objects/temple/dashButton00.png"

    if side == "left"
        Ahorn.drawSprite(ctx, texture, 20, 25, rot=pi)
    elseif side == "right"
        Ahorn.drawSprite(ctx, texture, 8, 8)
    elseif side == "up"
        Ahorn.drawSprite(ctx, texture, 9, 20, rot=-pi / 2)
    elseif side == "down"
        Ahorn.drawSprite(ctx, texture, 27, 7, rot=pi / 2)
    end
end

end

File: Loenn/entities/shroom_dash_switch.lua
local drawableSprite = require("structs.drawable_sprite")
local enums = require("consts.celeste_enums")
local utils = require("utils")

local shroom_dash_switch = {}

shroom_dash_switch.name = "ShroomHelper/ShroomDashSwitch"
shroom_dash_switch.depth = 0

shroom_dash_switch.fieldInformation = {
    windPatternOnCollision = {
        options = enums.wind_patterns,
        editable = false
    },
    side = {
        options = enums.dash_switch_sides,
        editable = false
    }
}

shroom_dash_switch.fieldOrder = { 
    "x", "y", "side", "windPatternOnCollision", "refillDashOnCollision", "doubleDashRefill", "isWindTrigger", "persistent"
}

shroom_dash_switch.placements = {
    {
        name = "shroom_dash_switch_up",
        data = {
            side = "Up",
            persistent = false,
            refillDashOnCollision = false,
            doubleDashRefill = false,
            isWindTrigger = false,
            windPatternOnCollision = "None"
        }
    },
    {
        name = "shroom_dash_switch_down",
        data = {
            side = "Down",
            persistent = false,
            refillDashOnCollision = false,
            doubleDashRefill = false,
            isWindTrigger = false,
            windPatternOnCollision = "None"
        }
    },
    {
        name = "shroom_dash_switch_left",
        data = {
            side = "Left",
            persistent = false,
            refillDashOnCollision = false,
            doubleDashRefill = false,
            isWindTrigger = false,
            windPatternOnCollision = "None"
        }
    },
    {
        name = "shroom_dash_switch_right",
        data = {
            side = "Right",
            persistent = false,
            refillDashOnCollision = false,
            doubleDashRefill = false,
            isWindTrigger = false,
            windPatternOnCollision = "None"
        }
    },
}

local doubleRefillTexture = "objects/sh_dashswitchdoublerefill/dashButtonMirror00"
local singleRefillTexture = "objects/sh_dashswitchrefill/dashButtonMirror00"
local noRefillTexture = "objects/sh_dashswitch/dashButtonMirror00"

local function getSwitchTexture(entity)
    local refill = entity.refillDashOnCollision
    local double = entity.doubleDashRefill

    if double and refill then
        return doubleRefillTexture
    elseif refill then
        return singleRefillTexture
    else
        return noRefillTexture
    end
end

function shroom_dash_switch.selection(room, entity)
    local side = entity.side

    if side == "Right" then
        return utils.rectangle(entity.x - 2, entity.y - 1, 12, 18)
    elseif side == "Left" then
        return utils.rectangle(entity.x - 2, entity.y - 1, 12, 18)
    elseif side == "Up" then
        return utils.rectangle(entity.x - 1, entity.y - 2, 18, 12)
    elseif side == "Down" then
        return utils.rectangle(entity.x - 1, entity.y - 2, 18, 12)
    end
end

function shroom_dash_switch.sprite(room, entity)
    local side = entity.side

    local switchSprite = drawableSprite.fromTexture(getSwitchTexture(entity), entity)

    if side == "Up" then
        switchSprite:setJustification(0.5, 1)
        switchSprite:addPosition(-16, 8)
        switchSprite.rotation = math.pi / 2
    elseif side == "Right" then
        switchSprite:setJustification(0, 0.5)
        switchSprite:addPosition(24, 8)
        switchSprite.rotation = math.pi
    elseif side == "Down" then
        switchSprite:setJustification(0.5, 0)
        switchSprite:addPosition(-16, 0)
        switchSprite.rotation = -math.pi / 2
    elseif side == "Left" then
        -- 
        switchSprite:setJustification(1, 0.5)
        switchSprite:addPosition(32, 8)
    end

    return switchSprite
end

return shroom_dash_switch
